### ISSUE TITLE
ContactQL should disallow ~ searching where value has fewer than 3 chars

### ISSUE THREAD

**Issue by rowanseymour** [nyaruka/goflow#842]
Since this can't be supported in ES

**Comment by nicpottier** [comment]
So looked at the code / tests for the indexer and it is slightly more complicated.

Name searches are always prefix searches across the first 8 characters.. so 2 character minimum length there but it needs to START with the letters (after tokenization). I think names like "Ed" lead us down this path, we want to match "Edward" there but not "Jared":
  https://github.com/nyaruka/rp-indexer/blob/master/indexer.go#L470

For URNs, it is a normal trigram substring:
  https://github.com/nyaruka/rp-indexer/blob/master/indexer.go#L454

So there substring match is fine, but the right side needs to be 3 characters or more.

**Comment by rowanseymour** [comment]
Is there a good reason those are different? Does seem a little weird that `~` in our query language can have a different meaning for `name` than everything else.

We can update goflow if that's the behavior we want but until now `name ~ X` has always just meant `name contains X`.

**Comment by nicpottier** [comment]
I think we started with them being the same, but the `ed` case below convinced us prefix makes more sense for names. So would say we should stick to that.

**Comment by rowanseymour** [comment]
What are the rules for `language` (and `timezone` eventually)?

**Comment by rowanseymour** [comment]
I also feel like the "Ed" example is bit language specific. If I were searching in Kinyarwanda I might want "zuba" to match "Akazuba", "Kazuba" etc.

**Comment by nicpottier** [comment]
Language is just a term query (ie, must match all of `eng`). 

No idea what we will do around timezone..

**Comment by rowanseymour** [comment]
I had no idea the goflow/python/ES implementations had ended up different in this way. Goflow's implementation has always been based around data type. In that regard `name`, `language`, `tel` are all just text and the operators have the same behavior. To summarize:

| Operator | Text | Number | Date
|--|--|--|--
| = | case-insensitive eq | eq | eq
| != | case-insensitive not eq | not eq | not eq
| ~ | case-insensitive contains | _not supported_  | _not supported_ 
| > >= < <= | _not supported_ | gt, gte, lt, lte | gt, gte, lt, lte

I think there's an argument to made for being simplest approach for users to grok.

**Comment by nicpottier** [comment]
We started with simplest and changed to prefix because we didn't feel the experience was right. Though they are slightly different I don't think we have received any complaints about it. (while we did with the substring case which is why we changed it)

**Comment by rowanseymour** [comment]
Frankly I can't remember that change and if we were discussing it today I'd suggest it's mistake as it's kinda language specific. But it sounds like you really don't want to consider changing it, so can we agree that `name` is the only exception? Can we make text fields, URNs, language, timezone all behave (and error) consistently?

**Comment by nicpottier** [comment]
Yes, realize it is language specific, though the majority of our customers are using Latin languages where that works well.

Are you talking about the `~` operator? There are performance reasons we can't turn that on for everything and we only have that on URNs. Same with `>`, `<` etc on text.

Or are you talking about something else?

**Comment by rowanseymour** [comment]
I'm talking about making Goflow in-code contact searches match the searches in ES. Please look at the table above and let's start figuring out all the ways in which they are different. Nowhere do we support `>` on text so that doesn't matter.

If `~` in ES only works for URNs and not contact fields then that's a significant difference that we need to address.

**Comment by nicpottier** [comment]
Apologies, I did look at the table, maybe I misunderstood.

`~` has never worked on text fields, that has always been the case as far as I know. IE:
  https://github.com/nyaruka/mailroom/blob/master/search/search.go#L160
  https://github.com/nyaruka/rapidpro/blob/master/temba/contacts/search/parser.py#L173

It does look like the Python evaluator had the same bug with name substring though:
  https://github.com/nyaruka/rapidpro/blob/master/temba/contacts/search/parser.py#L333

(that's a tough one because it is mostly determined by the index itself, not the query)

As far as I know that's the only thing that is different.

**Comment by nicpottier** [comment]
And let me fully explain my thought process on why I don't think we should change prefixing..

We made that decision deliberately when our heads were deep in the weeds of contact searching. We had tried a version of just trigrams but that wasn't right (it doesn't work for two letter names for one) and moving to bigrams didn't feel right either due to index bloat. So at that time we decided prefix was the right balance.

I don't remember all the details of that decision now, but I do know we spent a lot of time tuning behavior and indexes and we did it with a lot more knowledge and context than we currently have on this particular topic.

So my bias is to trust our past selves that made that decision, trust that they were informed on the pros and cons and familiarity with that problem more than our current selves which have only a glancing familiarity with all the cases there.

So basically that the burden of proof to change there is on the present us and I don't think just that it feels more consistent is a strong enough argument to overcome smarter past-us.

That's my thought process there.

**Comment by ericnewcomer** [comment]
The unwritten part to remember is that we are all getting dumber as we decay.

**Comment by rowanseymour** [comment]
We can keep using the ES version of searching by `name`. It's just a bit 🤦‍♂ that these changes didn't make it back into the Python or Goflow implementations of search and we let these things be different for so long. Good argument for trying to have as a few search implementations as we can.

If I'm understanding rp-indexer correctly, 

`~` for `name` needs to mean:
 - tokenize the input by whitespace
 - error if there are no tokens longer than 1 char
 - evaluation ignores any single character tokens
 - match if first up to 8 characters are same (this is bit weird cuz `"Alexandra" ~ "Alejandro"` is true)

`~` for a URN means:
- error if value less than 3 chars
- case insensitive contains

and it's not supported for anything else... an error I guess.

Currently it's an _evaluation_ error if you try to use `~` with something like a date or a number, but I think we should make these _parser_ errors.. no reason to not catch them earlier.

**Comment by nicpottier** [comment]
Yep, I think that's all right.

**Comment by rowanseymour** [comment]
Fixed in #843
